crates/ty_ide/src/goto_definition.rs

```diff
@@ -29,3 +29,598 @@ pub fn goto_definition(
+def my_function():
+    return "hello"
+"#,
+            )
+            .source(
+                "mymodule.pyi",
+                r#"
+def my_function(): ...
+"#,
+            )
+            .build();
+
+        assert_snapshot!(test.goto_definition(), @r"
+        info[goto-definition]: Definition
+         --> mymodule.pyi:1:1
+          |
+        1 |
+          | ^
+        2 | def my_function(): ...
+          |
+        info: Source
+         --> main.py:2:6
+          |
+        2 | from mymodule import my_function
+          |      ^^^^^^^^
+          |
+        ");
+    }
+
+    /// goto-definition on a module ref should go to the .py not the .pyi
+    #[test]
+    fn goto_definition_stub_map_module_ref() {
+        let test = CursorTest::builder()
+            .source(
+                "main.py",
+                "
+import mymodule
+x = mymo<CURSOR>dule
+",
+            )
+            .source(
+                "mymodule.py",
+                r#"
+def my_function():
+    return "hello"
+"#,
+            )
+            .source(
+                "mymodule.pyi",
+                r#"
+def my_function(): ...
+"#,
+            )
+            .build();
+
+        assert_snapshot!(test.goto_definition(), @r#"
+        info[goto-definition]: Definition
+         --> mymodule.py:1:1
+          |
+        1 |
+          | ^
+        2 | def my_function():
+        3 |     return "hello"
+          |
+        info: Source
+         --> main.py:3:5
+          |
+        2 | import mymodule
+        3 | x = mymodule
+          |     ^^^^^^^^
+          |
+        "#);
+    }
+
+    /// goto-definition on a function call should go to the .py not the .pyi
+    #[test]
+    fn goto_definition_stub_map_function() {
+        let test = CursorTest::builder()
+            .source(
+                "main.py",
+                "
+from mymodule import my_function
+print(my_func<CURSOR>tion())
+",
+            )
+            .source(
+                "mymodule.py",
+                r#"
+def my_function():
+    return "hello"
+
+def other_function():
+    return "other"
+"#,
+            )
+            .source(
+                "mymodule.pyi",
+                r#"
+def my_function(): ...
+
+def other_function(): ...
+"#,
+            )
+            .build();
+
+        assert_snapshot!(test.goto_definition(), @r#"
+        info[goto-definition]: Definition
+         --> mymodule.py:2:5
+          |
+        2 | def my_function():
+          |     ^^^^^^^^^^^
+        3 |     return "hello"
+          |
+        info: Source
+         --> main.py:3:7
+          |
+        2 | from mymodule import my_function
+        3 | print(my_function())
+          |       ^^^^^^^^^^^
+          |
+        "#);
+    }
+
+    /// goto-definition on a function that's redefined many times in the impl .py
+    ///
+    /// FIXME(?): Currently this yields all instances, presumably it should prefer the last one?
+    #[test]
+    fn goto_definition_stub_map_function_redefine() {
```

Comment:
Genuinely unsure what's correct here but I can definitely say for certain this behaviour naturally falls out of the fairly naive logic in my implementation.